### ISSUE TITLE
Address Already in Use Error When Training on 2 GPUS and Starting a New Job on the Remaining 2 GPUS

### ISSUE THREAD

**Issue by iprovalo** [coqui-ai/TTS#1678]
### Describe the bug

Following the steps on the [tutorial] (https://tts.readthedocs.io/en/latest/tutorial_for_nervous_beginners.html)

GlowTTS:
`python3 -m trainer.distribute --script train.py --gpus "0,1"`

Vocoder:
`python3 -m trainer.distribute --script train_vocoder.py --gpus "2,3"`

Getting this exception for the second command: `Address already in use`

### To Reproduce

1. Download the dataset per this [tutorial] (https://tts.readthedocs.io/en/latest/tutorial_for_nervous_beginners.html)

2. Run GlowTTS:
`python3 -m trainer.distribute --script train.py --gpus "0,1"`

3. Run Vocoder:
`python3 -m trainer.distribute --script train_vocoder.py --gpus "2,3"`

### Logs

```shell
Traceback (most recent call last):
  File "train_vocoder.py", line 44, in <module>
    TrainerArgs(), config, output_path, model=model, train_samples=train_samples, eval_samples=eval_samples
  File "/apps/tts/Trainer/trainer/trainer.py", line 460, in __init__
    self.config.distributed_url,
  File "/apps/tts/Trainer/trainer/utils/distributed.py", line 62, in init_distributed
    group_name=group_name,
  File "/usr/local/lib/python3.7/dist-packages/torch/distributed/distributed_c10d.py", line 595, in init_process_group
    store, rank, world_size = next(rendezvous_iterator)
  File "/usr/local/lib/python3.7/dist-packages/torch/distributed/rendezvous.py", line 186, in _tcp_rendezvous_handler
    store = _create_c10d_store(result.hostname, result.port, rank, world_size, timeout)
  File "/usr/local/lib/python3.7/dist-packages/torch/distributed/rendezvous.py", line 161, in _create_c10d_store
    hostname, port, world_size, start_daemon, timeout, multi_tenant=True
RuntimeError: The server socket has failed to listen on any local network address. The server socket has failed to bind to [::]:54321 (errno: 98 - Address already in use). The server socket has failed to bind to ?UNKNOWN? (errno: 98 - Address already in use).
```

### Environment

```shell
{
    "CUDA": {
        "GPU": [
            "NVIDIA A100-SXM4-40GB",
            "NVIDIA A100-SXM4-40GB",
            "NVIDIA A100-SXM4-40GB",
            "NVIDIA A100-SXM4-40GB"
        ],
        "available": true,
        "version": "11.5"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "1.11.0+cu115",
        "TTS": "0.7.0",
        "numpy": "1.21.6"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.7.13",
        "version": "#61~18.04.3-Ubuntu SMP Fri Oct 1 14:04:01 UTC 2021"
    }
}
```

### Additional context

First command starts with this log:

```
['train.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205837', '--use_ddp=true', '--rank=0']
['train.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205837', '--use_ddp=true', '--rank=1']
['train.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205837', '--use_ddp=true', '--rank=2']
['train.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205837', '--use_ddp=true', '--rank=3']
 > Using CUDA: True
 > Number of GPUs: 4
```

Second command starts with this log:

```
['train_vocoder.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205851', '--use_ddp=true', '--rank=0']
['train_vocoder.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205851', '--use_ddp=true', '--rank=1']
['train_vocoder.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205851', '--use_ddp=true', '--rank=2']
['train_vocoder.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_06_21-205851', '--use_ddp=true', '--rank=3']
 > Using CUDA: True
 > Number of GPUs: 4
[W socket.cpp:401] [c10d] The server socket has failed to bind to [::]:54321 (errno: 98 - Address already in use).
[W socket.cpp:401] [c10d] The server socket has failed to bind to ?UNKNOWN? (errno: 98 - Address already in use).
[E socket.cpp:435] [c10d] The server socket has failed to listen on any local network address.
```

**Comment by erogol** [comment]
you can use a different url overriding the related field in the config.

**Comment by iprovalo** [comment]
Thank you, @erogol .  Is it `distributed_url` parameter?

**Comment by erogol** [comment]
you can even override it on terminal by something like '--coqpit.distributed_url'

**Comment by iprovalo** [comment]
Thank you, @erogol , works great!

**Comment by iprovalo** [comment]
@erogol here is what happens when I use the distributed training with multiple GPUs.  At first everything looks OK, but then the second job fails.  I ran this on two language sets - `en` and `ru`.

A setup is 5 GPUS for the model, and 3 GPUS for the vocoder:
```
python -m trainer.distribute --script train_glow_tts_en.py --gpus "0,1,2,3,4"
python -m trainer.distribute --script train_hifigan_vocoder_en.py --gpus "5,6,7" --coqpit.distributed_url "tcp://localhost:54322"
```

Hifigan fails with no clear error, basically saying the run was removed:

Traceback (most recent call last):
  File "/apps/tts/Trainer/trainer/trainer.py", line 1492, in fit
    self._fit()
  File "/apps/tts/Trainer/trainer/trainer.py", line 1486, in _fit
    self.save_best_model()
  File "/apps/tts/Trainer/trainer/trainer.py", line 1576, in save_best_model
    save_func=self.dashboard_logger.save_model,
  File "/apps/tts/Trainer/trainer/io.py", line 186, in save_best_model
    fs.copy(checkpoint_path, shortcut_path)
  File "/usr/local/lib/python3.7/dist-packages/fsspec/spec.py", line 884, in copy
    self.cp_file(p1, p2, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/fsspec/implementations/local.py", line 121, in cp_file
    raise FileNotFoundError
FileNotFoundError

`ps -ef | grep python` and `nvidia-smi` both shows ghost process usages.

glow_tts just hangs for the `en-US` on the phoneme generation step (on a single GPU the same step takes 35min - 45K files in ljspeech)

Similar to `hifigan`, a lot of ghost processes are still showing on the `nvidia-smi` and python processes.

`ps -ef | grep python` for the `en` set also shows defunct processes right from the start.

I have not observed any of this on my single-GPU runs.

**Comment by Dapwner** [comment]
Hihi, I managed to get the very same error as @iprovalo when training a tacotron2 model (following an LJS recipe) on 2 gpus. The error does not appear with a single gpu.

Interestingly, I have tried to run the code on our nvidia dgx station (4 V100 gpus) with the following gpus:   0 + 1,     0 + 3,     1 + 3,     1 + 2
and it turned out that the error did not appear in the first two cases (when gpu 0 was included), but did occur in the latter two (when gpu 0 was not present).

When the error occurs, the code gets stuck in waiting, so a manual interrupt is necessary. After interrupting, the process on the second of the gpus is not removed and becomes a ghost process (occupying memory space, but 0% gpu usage)

**The error message**
```
 > BEST MODEL : /666/coquinew2/coqui/run-June-25-2022_08+27PM-577ec406/best_model_102.pth
  ! Run is removed from /666/coquinew2/coqui/run-June-25-2022_08+27PM-577ec406
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/trainer/trainer.py", line 1501, in fit
    self._fit()
  File "/opt/conda/lib/python3.7/site-packages/trainer/trainer.py", line 1495, in _fit
    self.save_best_model()
  File "/opt/conda/lib/python3.7/site-packages/trainer/trainer.py", line 1587, in save_best_model
    save_func=self.dashboard_logger.save_model,
  File "/opt/conda/lib/python3.7/site-packages/trainer/io.py", line 186, in save_best_model
    fs.copy(checkpoint_path, shortcut_path)
  File "/opt/conda/lib/python3.7/site-packages/fsspec/spec.py", line 884, in copy
    self.cp_file(p1, p2, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/fsspec/implementations/local.py", line 125, in cp_file
    raise FileNotFoundError
FileNotFoundError

```

**After Ctrl+C interrupt:**

```
^CTraceback (most recent call last):
  File "/opt/conda/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.7/site-packages/trainer/distribute.py", line 73, in <module>
    distribute()
  File "/opt/conda/lib/python3.7/site-packages/trainer/distribute.py", line 59, in distribute
    p.wait()
  File "/opt/conda/lib/python3.7/subprocess.py", line 1019, in wait
    return self._wait(timeout=timeout)
  File "/opt/conda/lib/python3.7/subprocess.py", line 1653, in _wait
    (pid, sts) = self._try_wait(0)
  File "/opt/conda/lib/python3.7/subprocess.py", line 1611, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt
```

I hope my feedback input is of some use.

**Comment by lexkoro** [comment]
I had the same thing happen, because of a long running task in the dataloader, so it times out and the whole process just hangs. Might be related to the phoneme preprocessing in this case.

**Comment by iprovalo** [comment]
@lexkoro I am not sure about the phoneme preprocessing in this case.  My second job was a vocoder, which starts training loops right away and fails on the first epoch save step.  On a keyboard interrupt event, it's the same message as seen by @Dapwner 

```
^CTraceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/apps/tts/Trainer/trainer/distribute.py", line 73, in <module>
    distribute()
  File "/apps/tts/Trainer/trainer/distribute.py", line 59, in distribute
    p.wait()
  File "/usr/lib/python3.7/subprocess.py", line 1019, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.7/subprocess.py", line 1653, in _wait
    (pid, sts) = self._try_wait(0)
  File "/usr/lib/python3.7/subprocess.py", line 1611, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt
```

**Comment by iprovalo** [comment]
@lexkoro I think you are right for the model side.  I re-ran the phoneme preprocessing on a single GPU, then restarted the process on multiple:

`nohup python3 -m trainer.distribute --script train_glow_tts_en_ljspeech.py --gpus "0,1,2,3,4" </dev/null > glow_tts_en_ljspeech.log 2>&1 &`

and it's going forward.  

I think there is a separate issue for the vocoder.  

I started on the same box on the remaining GPUs:

`
(root) /apps/tts/TTS # nohup python3 -m trainer.distribute --script train_hifigan_vocoder_en.py --gpus "5,6,7" --coqpit.distributed_url "tcp://localhost:54322" </dev/null > hifigan_en.log 2>&1 &
`

After a while I am getting this error and it hangs:

```
 > BEST MODEL : /apps/tts/data/output/run-July-05-2022_05+46AM-5c821d9f/best_model_137.pth
 ! Run is removed from /apps/tts/data/output/run-July-05-2022_05+46AM-5c821d9f
Traceback (most recent call last):
  File "/apps/tts/Trainer/trainer/trainer.py", line 1501, in fit
    self._fit()
  File "/apps/tts/Trainer/trainer/trainer.py", line 1495, in _fit
    self.save_best_model()
  File "/apps/tts/Trainer/trainer/trainer.py", line 1587, in save_best_model
    save_func=self.dashboard_logger.save_model,
  File "/apps/tts/Trainer/trainer/io.py", line 186, in save_best_model
    fs.copy(checkpoint_path, shortcut_path)
  File "/usr/local/lib/python3.7/dist-packages/fsspec/spec.py", line 884, in copy
    self.cp_file(p1, p2, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/fsspec/implementations/local.py", line 121, in cp_file
    raise FileNotFoundError
FileNotFoundError
```

`strace` shows the parent process is waiting on the child process, and the child process is doing some writing:

`ioctl(3, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffead048e30) = 0`

**Comment by lexkoro** [comment]
Guess the core issue is the same, a task takes to long and the dataloader times out?

**Comment by iprovalo** [comment]
I ended up debugging this more at runtime and tested a fix.  

It turns out that the current Trainer `distributed.py` behavior is using a GPU id for identifying the current device rank coming from one of the env vars.  I opened a [PR](https://github.com/coqui-ai/Trainer/pull/58) for the Trainer project.  The proposed behavior is using the `torch.distributed.get_rank()`

**Comment by erogol** [comment]
@iprovalo how does the rank cause the issue you are mentioning?

**Comment by iprovalo** [comment]
@erogol @Dapwner 

If you request GPU `0` in your list of GPUs, it will be used correctly, as in the case with my first command, but otherwise, it's the first GPU id in your list which becomes the master.  Then there is a mismatch between a GPU ID (GPU ID is being used in the current released code) and the rank (rank `0` is treated as the master).  This also explains the test @Dapwner ran:

> Interestingly, I have tried to run the code on our nvidia dgx station (4 V100 gpus) with the following gpus: 0 + 1, 0 + 3, 1 + 3, 1 + 2
and it turned out that the error did not appear in the first two cases (when gpu 0 was included), but did occur in the latter two (when gpu 0 was not present).

Basically, the exception happens on the distributed training with GPUs list starting at `1` and up.  Essentially, they run master-less.  No model files are being stored by the master, then this exception happens - `FileNotFoundError`.

So, in my second example I used GPU list `5,6,7`:

```/apps/tts/TTS # nohup python3 -m trainer.distribute --script train_hifigan_vocoder_en.py --gpus "5,6,7" --coqpit.distributed_url "tcp://localhost:54322" </dev/null > hifigan_en.log 2>&1 &```

A rank is printed correctly in the logs initially - GPU 5 gets rank 0.  But then during the training at runtime, once it is run on the GPUs, the pytorch gets GPU ID, instead of the rank using the current released code, so it looses the master all together, thus the ghost processes.

**Comment by erogol** [comment]
fixed by 👟 v0.0.14
